### PR TITLE
this commit allows to "hijack" the Serenity reporter and on top of it…

### PR DESCRIPTION
**Pull request by eitzenbe**
… perform additional reporting/forwarding of test run progress to external tools (we use this in tiger to show a browser UI tracing the current testrun in realtime)

see https://github.com/gematik/app-Tiger

### CODE REVIEW

**Comment by eitzenbe** [comment]
A screenie of the workflow UI can be found here: https://gematik.github.io/app-Tiger/Tiger-User-Manual.html#_workflow_ui

**Comment by wakaleo** [comment]
Interesting. Could you add some unit or smoke tests to demonstrate how it works?

**Comment by eitzenbe** [comment]
Sure, I see no tests at all for the SerenityReporter so far, but will get som ebasic stuff up'n running to show and verify the feature request ;)

**Comment by eitzenbe** [comment]
@wakaleo I have refactored the code and removed the callback feature but refactored the runner to have a base runner that is extendable and contains most code and also made the Reporter's major functionality reachable from an extending class. This doesn't change any functionality, so no need for unit tests there. I am attaching two files to illustrate how this can be used to extend the reporter plugin .... Any chance this could go in with the next release? 3.6.x
[TigerSerenityReportPlugin.java.txt](https://github.com/serenity-bdd/serenity-core/files/10560620/TigerSerenityReportPlugin.java.txt)
[TigerCucumberListener.java.txt](https://github.com/serenity-bdd/serenity-core/files/10560621/TigerCucumberListener.java.txt)

**Comment by wakaleo** [comment]
@cliviu could you take a look at this?

**Comment by eitzenbe** [comment]
@cliviu any timeline for review? We would need this for Tiger release and customers are waiting ;)

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+import net.serenitybdd.cucumber.suiteslicing.ScenarioFilter;
+import net.serenitybdd.cucumber.suiteslicing.TestStatistics;
+import net.serenitybdd.cucumber.suiteslicing.WeightedCucumberScenarios;
+import net.serenitybdd.cucumber.util.PathUtils;
+import net.serenitybdd.cucumber.util.Splitter;
+import net.thucydides.core.ThucydidesSystemProperty;
+import net.thucydides.core.environment.SystemEnvironmentVariables;
+import net.thucydides.core.steps.StepEventBus;
+import net.thucydides.core.util.EnvironmentVariables;
+import org.jetbrains.annotations.NotNull;
```
Comment (cliviu) [inline serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:41]: why the jetbrains @NotNull annotation?

Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:41]: my bad will be removed

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+import org.junit.runners.ParentRunner;
+import org.junit.runners.model.InitializationError;
+import org.junit.runners.model.RunnerScheduler;
+import org.junit.runners.model.Statement;
+
+@Slf4j
+public class CucumberSerenityBaseRunner extends ParentRunner<ParentRunner<?>> {
+
+
+    protected EventBus bus;
```
[the remark this answers is not in the document]
Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:54]: Would you mind to elaborate? in what aspect is package friendly a better approach here?

[the remark this answers is not in the document]
Comment (cliviu) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:53]: not package friendly but private

[the remark this answers is not in the document]
Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:53]: But we need access to these fields if we want to customize the runner, else the base class has no use at all...

[the remark this answers is not in the document]
Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:53]: I could try to minimize the accessability and make the fields private, getters protected and maybe I can circumvent having any PROTECTED setter method  (plugins and context fields are troublesome)

[the remark this answers is not in the document]
Comment (cliviu) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:53]: yes, this is what i mean. even if is arguable, i still find the protected fields a bad smell in code

**Comment by eitzenbe** [comment]
@cliviu new push, should resolve your remarks ;)

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java]: done see commit below ;)

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    private EventBus bus;
+
+    private Plugins plugins;
+
+    private List<ParentRunner<?>> children;
+
+    private List<Feature> features;
+    private CucumberExecutionContext context;
+
+    protected boolean multiThreadingAssumed = false;
```
Comment (cliviu) [inline serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:65]: still protected

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+
+
+    static ThreadLocal<RuntimeOptions> RUNTIME_OPTIONS = new ThreadLocal<>(); // NOSONAR
```
Comment (cliviu) [inline serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:68]: looks like this comment is not needed here

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+    private static RuntimeOptions DEFAULT_RUNTIME_OPTIONS; // NOSONAR
```
Comment (cliviu) [inline serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:70]: same here

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java (hunk 4 above)
Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java:68]: This is needed as Sonar code analysis is biting us here with a false positive warning

File: serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java (hunk 3 above)
Comment (eitzenbe) [reply serenity-cucumber/src/main/java/io/cucumber/junit/CucumberSerenityBaseRunner.java]: I think it was protected in the original code, svejedno I have changed it to private see new push

**Comment by eitzenbe** [comment]
@cliviu new push, should resolve also your second set of remarks ;)

**Pull request merged**